File: graphr/src/graphr/io/JsonVisitor.java
package graphr.io;

import graphr.data.GHT;
import graphr.data.JsonArrayState;
import graphr.data.JsonKeyValueState;
import graphr.data.PrimData;
import graphr.graph.Edge;
import graphr.graph.Edge.Direction;
import graphr.graph.Graph;
import graphr.graph.GraphData;
import graphr.graph.GraphDataVisitor;
import graphr.graph.GraphElementVisitor;
import graphr.graph.Vertex;

import java.util.Hashtable;
import java.util.Iterator;
import java.util.Map;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.json.JSONArray;
import org.json.JSONObject;

/**
 * Attempt to create a visitor pattern for graph serialization. <br>
 * Visitor that serializes/desirializes a graph into customized JSON output of
 * the following format:
 * 
 * <pre>
 * {
 *  "vertices": [
 * 	 {
 * 	  "type" : "Vertex"
 * 	  "id" : id
 *   "data" : { "key1" : "value1", "key2" : "value2", ...}
 *   "edges" : [{ "type": "Edge", "id" : id, "target" : targetId, "data": {same as vertex} }, {}, ...]
 * 	 },
 *  {...} 
 *  ], 
 *  "type" : graph
 * }
 * </pre>
 *
 * Assuming: https://github.com/douglascrockford/JSON-java
 *
 *
 *
 */
public class JsonVisitor<DV extends GraphData, DE extends GraphData> implements GraphElementVisitor {
	private static Logger log = LogManager.getLogger();
	private static final String EDGE_TARGET_KEY = "JSonVisitor_Edge_target";
	private static final String EDGE_SOURCE_KEY = "JSonVisitor_Edge_source";

	JsonKeyValueState root;
	JsonArrayState verticesForJson;
	GraphDataVisitor dataVisitor;

	@Override
	public void before() {
		dataVisitor = new GHTJsonVisitor();
	}

	@Override
	public void visit(Graph<?, ?> graph) {
		root = new JsonKeyValueState();
		root.add("type", "Graph");
		verticesForJson = new JsonArrayState();
	}

	@Override
	public void visit(Vertex<?, ?> vertex) {

		JsonKeyValueState vertexProp = new JsonKeyValueState();

		vertexProp.add("type", "Vertex");
		vertexProp.add("id", new Long(vertex.getId()).toString());
		vertexProp.add("data", vertex.getData() != null ? vertex.getData().accept(dataVisitor).toString() : "null");

		// get JSON array for edges
		JsonArrayState edgesForJson = new JsonArrayState();
		for (Edge<?, ?> e : vertex.getEdges()) {
                    //create only outgoing edges
                    if(e.getSource().getId() == vertex.getId()){
			JsonKeyValueState edgeProp = new JsonKeyValueState();
			edgeProp.add("type", "Edge");
			edgeProp.add("id", new Long(e.getId()).toString());
			edgeProp.add("data", (e.getData() == null ? "null" : e.getData().accept(dataVisitor).toString()));
			edgeProp.add("target", (e.getTarget() == null ? "null" : new Long(e.getTarget().getId()).toString()));
			edgesForJson.add(edgeProp.getAsJson());
                    }
		}
		vertexProp.add("edges", edgesForJson.getAsJson());

		// add serialized vertex into the list of vertices
		verticesForJson.add(vertexProp.getAsJson());
	}

	@Override
	public void visit(Edge<?, ?> edge) {
		// do nothing
	}

	@Override
	public void after() {
		root.add("vertices", verticesForJson.getAsJson());
	}

	public String getJsonString() {
		return root != null ? root.getAsJson() : "null";
	}

	/**
	 * Experimental method to parse graph in JSON. Initialization of edges is
	 * done in two steps. Firstly they are created during parsing a vertex but
	 * their target vertex is set to null. Once all vertices are parsed, proper
	 * references to target vertex (of the edge) is set. We use help structure
	 * that is map of edge to id of the vertex. Assuming:
	 * https://github.com/douglascrockford/JSON-java
	 * 
	 * @param readString
	 *            String to be parsed
	 * @return Initialized graph
	 */
	public Graph<GHT, GHT> parseJsonString(String readString) {

		JSONObject jo = new JSONObject(readString);

		// Init

		Graph<GHT, GHT> parsedGraph = new Graph<GHT, GHT>();
		if (!jo.getString("type").equals("Graph")) {
			throw new IllegalArgumentException("Graph expected.");
		}

		// Add vertices

		JSONArray vertices = jo.getJSONArray("vertices");
		int vertexCount = vertices.length();

		for (int i = 0; i < vertexCount; i++) {
			JSONObject vjo = vertices.getJSONObject(i);
			parsedGraph.addVertex(jSonObjectToVertex(vjo));
		}

		// Fix edge targets

		Hashtable<Long, Vertex<GHT, GHT>> vTable = parsedGraph.getVerticesAsHashtable();
		for (Vertex<GHT, GHT> v : vTable.values()) {
			Iterator<Edge<GHT, GHT>> edges = v.getEdges(Direction.BOTH).iterator();
			while (edges.hasNext()) {
				Edge<GHT, GHT> edge = edges.next();
				Hashtable<String, PrimData> edgeData = edge.getData().getTable();
				if (edgeData.containsKey(JsonVisitor.EDGE_TARGET_KEY)) {
					PrimData target = (PrimData) edgeData.get(JsonVisitor.EDGE_TARGET_KEY);
					Vertex<GHT, GHT> targetVertex = vTable.get(target.o());
					if (targetVertex != null) {
						edge.setTarget(targetVertex);
						edgeData.remove(JsonVisitor.EDGE_TARGET_KEY);

						// Add missing incoming edge to vertex
						if (targetVertex.getEdges(Direction.BOTH) == null
								|| !targetVertex.getEdges(Direction.BOTH).contains(edge)) {
							targetVertex.addEdge(edge);
						}
					}
				}
				if (edgeData.containsKey(JsonVisitor.EDGE_SOURCE_KEY)) {
					PrimData source = (PrimData) edgeData.get(JsonVisitor.EDGE_SOURCE_KEY);
					Vertex<GHT, GHT> sourceVertex = vTable.get(source.o());
					edge.setSource(sourceVertex);
					edgeData.remove(JsonVisitor.EDGE_SOURCE_KEY);
				}
				if (edge.getSource() == null || edge.getTarget() == null) {
					edges.remove();
				}
			}
		}

		return parsedGraph;
	}

	public Vertex<GHT, GHT> jSonObjectToVertex(JSONObject jo) {

		if (!jo.get("type").equals("Vertex")) {
			throw new IllegalArgumentException("Vertex expected.");
		}

		// Data directly attached to vertex

		Vertex<GHT, GHT> v = new Vertex<GHT, GHT>(new GHT());
		long vertexId = jo.getLong("id");
		v.setId(vertexId);

		Object o = jo.get("data");
		if (o instanceof JSONObject) {
			JSONObject dataO = (JSONObject) o;
			v.setData(jSonObjectToGHT(dataO));
		} else {
			v.setData(new GHT());
		}

		// Edges

		Object edges = jo.get("edges");
		if (edges instanceof JSONArray) {

			JSONArray edgesArray = (JSONArray) edges;

			for (int i = 0; i < edgesArray.length(); i++) {
				JSONObject ejo = edgesArray.getJSONObject(i);
				v.addEdge(jSonObjectToEdge(ejo, vertexId));
			}
		}

		return v;
	}

	public Edge<GHT, GHT> jSonObjectToEdge(JSONObject jo, long sourceLong) {

		Edge<GHT, GHT> edge = new Edge<GHT, GHT>(new GHT());
		edge.setId(jo.getLong("id"));

		Object o = jo.get("data");
		if (o instanceof JSONObject) {
			JSONObject dataO = (JSONObject) o;
			edge.setData(jSonObjectToGHT(dataO));
		} else {
			edge.setData(new GHT());
		}

		Long targetLong = jo.getLong("target");
		edge.getData().put(JsonVisitor.EDGE_TARGET_KEY, new PrimData(targetLong));
		edge.getData().put(JsonVisitor.EDGE_SOURCE_KEY, new PrimData(sourceLong));

		return edge;
	}

	public GHT jSonObjectToGHT(JSONObject jo) {

		Iterator<String> keys = jo.keys();

		GHT table = new GHT();

		while (keys.hasNext()) {

			String key = keys.next();

			Object value = jo.get(key);
			table.put(key, new PrimData(value));

		}

		return table;
	}

	/**
	 * Helper method to parse given JSON map into the Graph hashtable
	 * 
	 * @param dataMap
	 *            JSON map of keys and properties
	 * @return Graph hashtable
	 */
	public GHT parseGHT(Map<?, ?> dataMap) {
		GHT data = new GHT();
		for (Object mapKey : dataMap.keySet()) {
			Object mapValue = dataMap.get(mapKey);
			log.debug("key: " + mapKey + "; value: " + mapValue);
			data.put(mapKey.toString(), mapValue.toString());
		}

		return data;
	}

}

File: graphrs/graphrs/io/JsonVisitor.swift
//
//  JsonVisitor.swift
//  graphrs
//
//  Created by Marcus Sellmann on 01.04.16.
//  Copyright © 2016 CITBDA. All rights reserved.
//

public enum JsonVisitorError: ErrorType {
    case GraphExpectedError
    case VertexExpectedError
    case EdgeExpectedError
    case GraphConversionError
}

/**
 * Attempt to create a visitor pattern for graph serialization. <br>
 * Visitor that serializes/desirializes a graph into customized JSON output of
 * the following format:
 *
 * <pre>
 * {
 *  "vertices": [
 * 	 {
 * 	  "type" : "Vertex"
 * 	  "id" : id
 *   "data" : { "key1" : "value1", "key2" : "value2", ...}
 *   "edges" : [{ "type": "Edge", "id" : id, "target" : targetId, "data": {same as vertex} }, {}, ...]
 * 	 },
 *  {...}
 *  ],
 *  "type" : graph
 * }
 * </pre>
 *
 * Assuming: https://github.com/douglascrockford/JSON-java
 *
 *
 *
 */
public class JsonVisitor: GraphElementVisitor {
    private static let EDGE_TARGET_KEY: String = "JSonVisitor_Edge_target"
    private static let EDGE_SOURCE_KEY: String = "JSonVisitor_Edge_source"
    
    var root = [String : String]()
    var verticesForJson = [String]()
    var dataVisitor: GraphDataVisitor? = nil
    
    func before() {
        dataVisitor = DictionaryElementDataJsonVisitor()
    }
    
    func visit(graph: Graph) {
        root["type"] = "Graph"
    }
    
    func visit(vertex: Vertex) {
        var vertexProp = [String : String]()
        
        vertexProp["type"] = "Vertex"
        vertexProp["id"] = String(vertex.id)
        vertexProp["data"] = (vertex.data as! DictionaryElementData).accept(dataVisitor!)
        
        // get JSON array for edges
        var edgesForJson = [String]()
        
        for e in vertex.getEdges()! {
            //create only outgoing edges
            if e.sideA.id == vertex.id {
                var edgeProp = [String : String]()
                edgeProp["type"] = "Edge"
                edgeProp["id"] = String(e.id)
                edgeProp["data"] = (e.data as! DictionaryElementData).accept(dataVisitor!)
                edgeProp["target"] = String(e.sideB.id)
                edgesForJson.append(JsonConverter.dictToJson(edgeProp))
            }
        }
        
        vertexProp["edges"] = JsonConverter.arrayToJson(edgesForJson)
        
        // add serialized vertex into the list of vertices
        verticesForJson.append(JsonConverter.dictToJson(vertexProp))
    }
    
    func visit(edge: Edge) {
        // do nothing
    }
    
    func after() {
        root["vertices"] = JsonConverter.arrayToJson(verticesForJson)
    }
    
    func getJsonString() -> String {
        return JsonConverter.dictToJson(root)
    }
    
    /**
     * Experimental method to parse graph in JSON. Initialization of edges is
     * done in two steps. Firstly they are created during parsing a vertex but
     * their target vertex is set to null. Once all vertices are parsed, proper
     * references to target vertex (of the edge) is set. We use help structure
     * that is map of edge to id of the vertex. Assuming:
     * https://github.com/douglascrockford/JSON-java
     *
     * @param readString
     *            String to be parsed
     * @return Initialized graph
     *//*
    func parseJsonString(readString: String) -> Graph {
        JSONObject jo = new JSONObject(readString);
        return parseJsonObject(jo);
    }
    
    public Graph<GHT, GHT> parseJsonStream(final InputStream stream) throws IllegalArgumentException {
        try {
            JSONTokener tokener = new JSONTokener(stream);
            JSONObject js = new JSONObject(tokener);
            return parseJsonObject(js);
        } catch (JSONException e) {
            throw new IllegalArgumentException(e);
        }
    }*/
    
    /**
     * Experimental method to parse graph in JSON. Initialization of edges is
     * done in two steps. Firstly they are created during parsing a vertex but
     * their target vertex is set to null. Once all vertices are parsed, proper
     * references to target vertex (of the edge) is set. We use help structure
     * that is map of edge to id of the vertex. Assuming:
     * https://github.com/douglascrockford/JSON-java
     *
     * @param jo
     *            JSONObject to be parsed
     * @return Initialized graph
     */
    func parseJsonObject(jo: String) throws -> Graph {
        do {
            // Init
            
            let parsedGraph = Graph()
            
            let data = try NSJSONSerialization.JSONObjectWithData(jo.dataUsingEncoding(NSUTF8StringEncoding)!, options: .AllowFragments)
            var rootDict = data as! [String : AnyObject]
            
            if (rootDict["type"] as! String != "Graph") {
                throw JsonVisitorError.GraphExpectedError
            }
            
            // Add vertices
            
            let verticesDict = rootDict["vertices"] as! [AnyObject]
            
            for vertex in verticesDict {
                let vert = vertex as! [String : AnyObject]
                
                if (vert["type"] as! String != "Vertex") {
                    throw JsonVisitorError.VertexExpectedError
                }
                
                do {
                    let v = try jsonObjectToVertex(vert)
                    parsedGraph.addVertex(v.id, vertex: v)
                } catch {}
            }
            
            // Fix edge targets
            
            let vertices = parsedGraph.getVertexDict()
            
            for v in vertices {
                let edges: [Edge] = v.1.getEdges(Direction.Undefined)!
                
                for edge in edges {
                    let edgeData = edge.data as! DictionaryElementData
                    
                    if let targetId = edgeData.d[JsonVisitor.EDGE_TARGET_KEY] {
                        if let targetV = vertices[try targetId.l()] {
                            edge.sideB = targetV
                            
                            // Add missing incoming edge to vertex
                            if (targetV.getEdges(Direction.Undefined)!.isEmpty || !(targetV.getEdges(Direction.Undefined)!.contains(edge))) {
                                targetV.addEdge(edge)
                            }
                        }
                    }
                    
                    if let sourceId = edgeData.d[JsonVisitor.EDGE_SOURCE_KEY] {
                        if let sourceV = vertices[try sourceId.l()] {
                            edge.sideA = sourceV
                        }
                    }
                    
                    if (edgeData.d[JsonVisitor.EDGE_SOURCE_KEY] == nil || edgeData.d[JsonVisitor.EDGE_TARGET_KEY] == nil) {
                        v.1.removeEdgeOnBothSides(edge)
                    }
                    
                    edgeData.removeData(JsonVisitor.EDGE_SOURCE_KEY)
                    edgeData.removeData(JsonVisitor.EDGE_TARGET_KEY)
                }
            }
            
            return parsedGraph
        } catch {
            throw JsonVisitorError.GraphConversionError
        }
    }
    
    public func jsonObjectToVertex(jo: [String : AnyObject]) throws -> Vertex {
        if (jo["type"] as! String != "Vertex") {
            throw JsonVisitorError.VertexExpectedError
        }
        
        // Data directly attached to vertex
        
        let v = Vertex()
        v.id = UInt64(String(jo["id"]!))!
        
        let vData = jo["data"] as! [String : AnyObject]
        v.data = jsonObjectToDictionaryElementData(vData)
        
        // Edges
        
        if let edges: [AnyObject]? = jo["edges"]! as? [AnyObject] {
            if (edges != nil) {
                for edge in edges! {
                    let eData = edge as! [String : AnyObject]
                    v.addEdge(try jsonObjectToEdge(eData, source: v))
                }
            }
        }
        
        return v
    }
    
    func jsonObjectToEdge(jo: [String : AnyObject], source: Vertex) throws -> Edge {
        if (jo["type"] as! String != "Edge") {
            throw JsonVisitorError.EdgeExpectedError
        }
        
        let eId = UInt64(String(jo["id"]!))!
        let targetId = UInt64(String(jo["target"]!))!
        let data = jo["data"] as! [String : AnyObject]
        
        let dictData = jsonObjectToDictionaryElementData(data)
        let edge: Edge = try Edge(sideA: source, sideB: source, direction: .Forward, data: dictData)
        edge.id = eId
        
        (edge.data as! DictionaryElementData).updateData(JsonVisitor.EDGE_SOURCE_KEY, value: PrimitiveData(l: source.id))
        (edge.data as! DictionaryElementData).updateData(JsonVisitor.EDGE_TARGET_KEY, value: PrimitiveData(l: targetId))
        
        return edge
    }
    
    func jsonObjectToDictionaryElementData(jo: [String : AnyObject]) -> DictionaryElementData {
        let dict = DictionaryElementData()
        
        for key in jo.keys {
            do {
                dict.updateData(key, value: try PrimitiveData(o: jo[key]!))
            } catch {}
        }
        
        return dict
    }
}
